source/hash_table.c

```diff
@@ -0,0 +1,582 @@
+
+    return state;
+}
+
+// Computes the correct size and max_load based on a requested size.
+static int update_template_size(struct hash_table_state *template, size_t expected_elements) {
+    size_t min_size = expected_elements;
+
+    if (min_size < 2) {
+        min_size = 2;
+    }
+
+    size_t mask = ~(size_t)0, size = 1;
+    while (size < min_size) {
+        size = size << 1;
+        mask = mask << 1;
+
+        if (size == 0) {
+            // Overflow
+            return AWS_ERROR_OOM;
+        }
+    }
+    mask = ~mask;
+
+    // Cross-check - make sure we didn't just overflow somehow.
+    if (size < expected_elements) {
+        return AWS_ERROR_OOM;
+    }
+
+    template->size = size;
+    template->max_load = template->max_load_factor * template->size;
+    if (template->max_load >= size) {
+        template->max_load = size - 1;
+    }
+
+    // Make sure we don't overflow when computing memory requirements either
+    size_t required_mem = aws_common_mul_size_saturating(template->size, sizeof(struct hash_table_entry));
+    if (required_mem == SIZE_MAX || (required_mem + sizeof(struct hash_table_state)) < required_mem) {
+        return AWS_ERROR_OOM;
+    }
+
+    template->size = size;
+    template->mask = mask;
+
+    return AWS_ERROR_SUCCESS;
+}
+
+int aws_common_hash_table_init(struct aws_common_hash_table *map,
+    struct aws_allocator *alloc, size_t size,
+    aws_hash_fn_t hash_fn, aws_equals_fn_t equals_fn, aws_hash_element_destroy_t destroy_fn) {
+
+    struct hash_table_state template;
+    template.hash_fn = hash_fn;
+    template.equals_fn = equals_fn;
+    template.destroy_fn = destroy_fn;
+    template.alloc = alloc;
+
+    template.entry_count = 0;
+    template.max_load_factor = 0.95; // TODO - make configurable?
+
+    update_template_size(&template, size);
+    map->pImpl = alloc_state(&template);
+
+    if (!map->pImpl) {
+        return AWS_ERROR_OOM;
+    }
+
+    return AWS_ERROR_SUCCESS;
+}
+
+void aws_common_hash_table_clean_up(struct aws_common_hash_table *map) {
+    aws_common_hash_table_clear(map);
+    aws_mem_release(((struct hash_table_state *)map->pImpl)->alloc, map->pImpl);
+}
+
+// Tries to find where the requested key is or where it should go if put.
+// Returns AWS_ERROR_SUCCESS if the item existed (leaving it in *entry),
+// or AWS_ERROR_HASHTBL_ITEM_NOT_FOUND if it did not (putting its destination
+// in *entry). Note that this does not take care of displacing whatever was in that entry before.
+
+// probe_idx is set to the probe index of the entry.
+
+static int find_entry1(struct hash_table_state *state, uint64_t hash_code, const void *key, struct hash_table_entry **p_entry, int *p_probe_idx);
+
+// Inlined fast path: Check the first slot, only.
+static int find_entry(struct hash_table_state *state, uint64_t hash_code, const void *key, struct hash_table_entry **p_entry, int *p_probe_idx) __attribute__((always_inline));
```

Comment:
this will only compile on GCC and CLang, let's add a macro definition that handles msvc as well in Common.h